src/systems/filecoin_blockchain/storage_power_consensus/storage_power_actor.go

```diff
@@ -56,21 +304,21 @@ func (spa *StoragePowerActor_I) Surprise(ticket block.Ticket) []addr.Address {
 	// var currBlockHeight util.UInt
 
 	// The number of miners that are challenged at this block
-	numSurprised := uint(len(spa.Miners())) / provingPeriod
+	numSurprised := uint(len(spa.PowerTable())) / provingPeriod
 
-	// TODO: seem inefficient but spa.Miners() is now a map from address to power
-	minerAddresses := make([]addr.Address, len(spa.Miners()))
+	// TODO: seem inefficient but spa.PowerTable() is now a map from address to power
+	minerAddresses := make([]addr.Address, len(spa.PowerTable()))
 
 	index := 0
-	for address := range spa.Miners() {
+	for address, _ := range spa.PowerTable() {
```

Comment:
i think this is going to pick all of them -- iterating over all of the power table. you want:

```go
table := spa.PowerTable()
numMiners := len(table)
random := 4 // TODO: get randomness from chain
offset := random % numMiners
for i := 0; i < newSurprised; i++ {
  j := (offset + i) % numMiners 
  minerAddresses[i] = table[j]
}
```
Note somewhere:
- this picks `newSurprised` consecutive miners, starting from a random `offset`
- good for sampling w/o replacement, but may be bad as it would be the same cohort always challenged at the same time
- if we want to pick properly random set, can: 
  - make a list of indices first, shuffle it, then pick consecutive from those 
  - or pick at random every time, but skip ones we've seen already (sample w/ replacement, but skip doubles)

Reply:
adding these as comments to Surprise, it should be its own PR. cc @nicola